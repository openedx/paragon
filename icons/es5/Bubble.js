function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
import * as React from "react";
var SvgBubble = function SvgBubble(props) {
  return /*#__PURE__*/React.createElement("svg", _extends({
    xmlns: "http://www.w3.org/2000/svg",
    width: 24,
    height: 24,
    viewBox: "0 -960 960 960",
    fill: "none"
  }, props), /*#__PURE__*/React.createElement("path", {
    d: "M480-840q75 0 140.5 28T735-735q49 49 77 114.5T840-480q0 75-28 140.5T735-225q-49 49-114.5 77T480-120q-49 0-98-10.5T296-170l43-43q24 17 63.5 25t77.5 8q127 0 213.5-86.5T780-480q0-127-86.5-213.5T480-780q-127 0-213.5 86.5T180-480q0 39 7.5 72.5T211-339l-45 45q-20-32-33-81t-13-105q0-75 28-140.5T225-735q49-49 114.5-77T480-840Zm55 519v-171L163-120l-43-43 372-372H321v-60h274v274h-60Z",
    fill: "currentColor"
  }));
};
export default SvgBubble;