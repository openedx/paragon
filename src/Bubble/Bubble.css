/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:19 GMT
 */

:root {
  --pgn-spacing-bubble-expandable-padding-x: .25rem;
  --pgn-spacing-bubble-expandable-padding-y: 0;
}
