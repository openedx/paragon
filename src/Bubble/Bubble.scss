@import "variables";

.pgn__bubble {
  line-height: normal;
  min-height: 1.5rem;
  min-width: 1.5rem;
  border-radius: 50%;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  font-size: $font-size-xs;

  @each $name, $colors in $bubble-variants {
    &.pgn__bubble-#{$name} {
      background-color: map-get($colors, "background-color");
      color: map-get($colors, "color");
    }
  }

  &:disabled,
  &.disabled {
    background: $gray-500;
  }

  &.expandable {
    padding: $bubble-expandable-padding-y $bubble-expandable-padding-x;
  }
}
