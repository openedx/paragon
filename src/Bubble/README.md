---
title: 'Bubble'
type: 'component'
components:
- Bubble
status: 'New'
categories:
- Status & metadata
designStatus: 'Done'
devStatus: 'Done'
notes: |
---

Represents the filled circle with a number of an icon, supporting the usual color variants

## Basic Usage

```jsx live
() => {
  return (
    <>
      <Stack direction="horizontal" gap={4} className="mb-2">
        <Bubble>
          1
        </Bubble>
        <Bubble variant="error">
          1
        </Bubble>
        <Bubble variant="warning">
          1
        </Bubble>
        <Bubble variant="success">
          1
        </Bubble>
      </Stack>
      <Stack direction="horizontal" gap={4} className="mb-2">
        <Bubble>
          <Icon src={Check} />
        </Bubble>
        <Bubble variant="error">
          <Icon src={Check} />
        </Bubble>
        <Bubble variant="warning">
          <Icon src={Check} />
        </Bubble>
        <Bubble variant="success">
          <Icon src={Check} />
        </Bubble>
      </Stack>
    </>
  );
}
```

### Disabled variant

```jsx live
() => {
  return (
    <Stack direction="horizontal" gap={4} className="mb-2">
      <Bubble disabled>
        1
      </Bubble>
      <Bubble disabled>
        <Icon src={Check} />
      </Bubble>
    </Stack>
  );
}
```

### Expandable

In the case of long content use `expandable` prop. It adds padding to `Bubble`. Padding value is configurable through `scss` variables.

```jsx live
() => {
  {/* start example state */}
  const [isExpandable, setIsExpandable] = useState(true);
  {/* end example state */}

  return (
    <>
      {/* start example form block */}
      <ExamplePropsForm
        inputs={[
          { value: isExpandable, setValue: () => setIsExpandable(!isExpandable), name: 'isExpandable' },
        ]}
      />
      {/* end example form block */}
      <Bubble variant="error" expandable={isExpandable}>
        1234
      </Bubble>
    </>
  );
}
```
