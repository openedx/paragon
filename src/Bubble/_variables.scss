$bubble-variants: (
  "success":    ( "background-color": $success, "color": $white ),
  "warning":    ( "background-color": $warning, "color": $white ),
  "error":      ( "background-color": $danger, "color": $white ),
  "primary":    ( "background-color": $primary, "color": $white ),
) !default;
$bubble-expandable-padding-x:              .25rem !default;
$bubble-expandable-padding-y:              0 !default;
