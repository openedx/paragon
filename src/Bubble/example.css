.pgn__bubble {
    background: purple !important;
}
