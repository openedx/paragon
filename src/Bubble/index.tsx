import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const STYLE_VARIANTS = ['primary', 'success', 'error', 'warning'] as const;

export type BubbleVariant = typeof STYLE_VARIANTS[number];

export interface BubbleProps {
  children: React.ReactNode;
  variant?: BubbleVariant;
  disabled?: boolean;
  className?: string;
  expandable?: boolean;
}

const Bubble = React.forwardRef<HTMLDivElement, BubbleProps>(({
  variant,
  className,
  children,
  disabled,
  expandable,
  ...props
}, ref) => (
  <div
    ref={ref}
    className={classNames(
      'pgn__bubble',
      `pgn__bubble-${variant}`,
      className,
      { disabled, expandable },
    )}
    {...props}
  >
    {children}
  </div>
));

Bubble.propTypes = {
  /** Specifies contents of the component. */
  children: PropTypes.node,
  /** The `Bubble` style variant to use. */
  variant: PropTypes.oneOf(STYLE_VARIANTS),
  /** Activates disabled variant. */
  disabled: PropTypes.bool,
  /** A class name to append to the base element. */
  className: PropTypes.string,
  /** Specifies whether to add padding to the `Bubble` or not. */
  expandable: PropTypes.bool,
};

Bubble.defaultProps = {
  children: null,
  variant: 'primary',
  disabled: false,
  className: undefined,
  expandable: false,
};

export default Bubble;
