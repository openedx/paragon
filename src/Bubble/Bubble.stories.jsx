import Bubble from './index';

export default {
  title: 'Bubble',
  component: Bubble,
  argTypes: {
    variant: {
      control: {
        type: 'select',
      },
    },
  },
};

const Template = (args) => <Bubble {...args} />;

export const Primary = Template.bind({});
Primary.args = { children: 'Primary', variant: 'primary' };

export const Success = Template.bind({});
Success.args = { children: 'Success', variant: 'success' };

export const Error = Template.bind({});
Error.args = { children: 'Error', variant: 'error' };

export const Warning = Template.bind({});
Warning.args = { children: 'Warning', variant: 'warning' };
