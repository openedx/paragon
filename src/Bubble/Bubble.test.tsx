import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Bubble from './index';

describe('<Bubble />', () => {
  describe('correct rendering', () => {
    it('successfully renders', () => {
      const tree = renderer.create(<Bubble>1</Bubble>).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with variant', () => {
      const wrapper = mount(<Bubble variant="error">1</Bubble>);
      const bubble = wrapper.find('.pgn__bubble');
      expect(bubble.hasClass('pgn__bubble-error')).toEqual(true);
    });
    it('renders with default variant', () => {
      const wrapper = mount(<Bubble>1</Bubble>);
      const bubble = wrapper.find('.pgn__bubble');
      expect(bubble.hasClass('pgn__bubble-primary')).toEqual(true);
    });
    it('renders with disabled variant', () => {
      const wrapper = mount(<Bubble disabled>1</Bubble>);
      const bubble = wrapper.find('.pgn__bubble');
      expect(bubble.hasClass('disabled')).toEqual(true);
    });
    it('renders with disabled variant', () => {
      const wrapper = mount(<Bubble expandable>1</Bubble>);
      const bubble = wrapper.find('.pgn__bubble');
      expect(bubble.hasClass('expandable')).toEqual(true);
    });
  });
});
