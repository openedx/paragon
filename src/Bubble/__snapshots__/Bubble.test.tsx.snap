// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Bubble /> correct rendering successfully renders 1`] = `
<div
  className="pgn__bubble pgn__bubble-primary"
>
  1
</div>
`;
